radicle/src/node.rs

```diff
@@ -1,32 +1,79 @@
 mod features;
 
-use amplify::WrapperMut;
-use std::io::{BufRead, BufReader, Write};
-use std::ops::Deref;
+use std::io::{BufRead, BufReader};
 use std::os::unix::net::UnixStream;
 use std::path::{Path, PathBuf};
-use std::{io, net};
+use std::{fmt, io, net};
 
+use amplify::WrapperMut;
 use crossbeam_channel as chan;
 use cyphernet::addr::{HostName, NetAddr};
-use nonempty::NonEmpty;
+use serde::de::DeserializeOwned;
+use serde::{Deserialize, Serialize};
+use serde_json as json;
 
 use crate::crypto::PublicKey;
-use crate::git;
 use crate::identity::Id;
-use crate::storage;
-use crate::storage::{Namespaces, RefUpdate};
+use crate::storage::RefUpdate;
 
 pub use features::Features;
 
 /// Default name for control socket file.
 pub const DEFAULT_SOCKET_NAME: &str = "radicle.sock";
 /// Default radicle protocol port.
 pub const DEFAULT_PORT: u16 = 8776;
-/// Response on node socket indicating that a command was carried out successfully.
-pub const RESPONSE_OK: &str = "ok";
-/// Response on node socket indicating that a command had no effect.
-pub const RESPONSE_NOOP: &str = "noop";
+
+/// Result of a command, on the node control socket.
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+#[serde(tag = "status")]
+pub enum CommandResult {
+    /// Response on node socket indicating that a command was carried out successfully.
+    #[serde(rename = "ok")]
+    Okay {
+        /// Whether the command had any effect.
+        #[serde(default, skip_serializing_if = "crate::serde_ext::is_default")]
+        updated: bool,
+    },
+    /// Response on node socket indicating that an error occured.
+    Error {
+        /// The reason for the error.
+        reason: String,
+    },
+}
+
+impl CommandResult {
+    /// Create an "updated" response.
+    pub fn updated() -> Self {
+        Self::Okay { updated: true }
+    }
+
+    /// Create an "ok" response.
+    pub fn ok() -> Self {
+        Self::Okay { updated: false }
+    }
```

Comment:
Are the 'updated' flags for user UI? Or something else?

If we run the command twice in a row, would it matter if the first is Ok(update: true) and the second is Ok(updated: false)?

Reply:
It's for user feedback yeah. User needs to know if something happened, or if it was a no-op.

> If we run the command twice in a row, would it matter if the first is Ok(update: true) and the second is Ok(updated: false)?

This would be the expected behavior for certain commands, yes.